### PR TITLE
pull request workflow

### PR DESCRIPTION
**Pull request by iamer**
Workflow for building and testing realm-js on pull requests

## ☑️ ToDos
* [ ] Check prebuilds are used correctly for all platforms
* [ ] Add BAAS test job
* [x] Check electron tests js local dependency link
* [x] enable all builds in the job matrix

### CODE REVIEW

File: .github/workflows/pr-realm-js.yml [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    steps:
+      # This will retrieve the realm-js github repository
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          submodules: "recursive"
+
+      # This sets node up, and can also pick the version if necessary
+      - name: Setup node version
+        uses: actions/setup-node@v1
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:33]: I think we would want to pin this version.

File: .github/workflows/pr-realm-js.yml [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+          #- { os: windows, runner: windows-latest, arch: x64, test-node: true, test-electron: true }
+          #- { os: windows, runner: windows-latest, arch: ia32 }
+          #- { os: darwin, runner: macos-latest, arch: x64, test-node: true, test-electron: true, test-react: true }
+          #- { os: ios, runner: macos-latest, arch: x64}
+          - { os: android, runner: ubuntu-latest, arch: x86_64 }
+          - { os: android, runner: ubuntu-latest, arch: armeabi-v7a }
+          - { os: android, runner: ubuntu-latest, arch: arm64-v8a }
+          - { os: android, runner: ubuntu-latest, arch: x86 }
+    steps:
+      # This will retrieve the realm-js github repository
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:25]:
```suggestion
```

File: .github/workflows/pr-realm-js.yml [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        run: |
+          Invoke-WebRequest -Uri https://static.realm.io/downloads/vcpkg.zip -OutFile C:\vcpkg.zip
+          Expand-Archive -Path C:\vcpkg.zip -DestinationPath C:\
+        shell: powershell
+
+      # ninja-build is used by default if available and results in faster build times
+      # In linux, electron requires a connected display.  We fake this by giving it a headless environment using xvfb
+      # Relevant issue: https://github.com/juliangruber/browser-run/issues/147
+      - name: Linux Environment setup
+        if: ${{ (matrix.variant.os == 'linux') && (steps.cache-prebuild.outputs.cache-hit != 'true') }}
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:61]: Why is `xvfb` not needed if the cache is hit?

File: .github/workflows/pr-realm-js.yml [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+          cache-name: cache-ios-build
+        with:
+          path: react-native/ios/build
+          key: build-ios-${{ hashFiles('./src/**') }}
+          restore-keys: |
+            build-ios-
+
+      - name: Setup environment
+        run: |
+          brew install watchman
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:159]: I'd suggest sticking with the single line variant for these one-liners.

```suggestion
        run: brew install watchman
```

File: .github/workflows/pr-realm-js.yml [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    runs-on: ubuntu-latest
+    name: Cleanup realm js
+    needs: [build-and-test, baas-realm-js]
+    if: always()
+    timeout-minutes: 10
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+        with:
+          submodules: false
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:299]: I'm wondering why this checkout is needed?

File: .github/workflows/pr-realm-js.yml [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+    if: ${{false}}
+    needs: [build-and-test, test-react-native-ios, test-react-native-android]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Cache js build artifacts
+        uses: actions/cache@v2
+        env:
+          cache-name: cache-js-build-artifacts
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:236]: Is there a good reason this set as an environment variable? I suppose it's to keep the `key` and `restore-keys` DRY?

File: .github/workflows/pr-realm-js.yml [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        with:
+          # npm cache files are stored in `~/.npm` on Linux/macOS
+          path: ~/.npm
+          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{hashFiles('**/package-lock.json')}}
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:240]: Feels redundant to have the `-build-` when we also have the `cache-name`.
```suggestion
          key: ${{ runner.os }}-${{ env.cache-name }}-${{hashFiles('**/package-lock.json')}}
```

File: .github/workflows/pr-realm-js.yml [hunk 8]
```diff
@@ -0,0 +1,306 @@
+name: pr-realm-js
+on: pull_request
+
+jobs:
+  build-and-test:
+    name: Build and test for ${{ matrix.variant.os }} ${{ matrix.variant.arch }}
+    runs-on: ${{ matrix.variant.runner }}
+    env:
+      npm_config_realm_local_prebuilds: ${{github.workspace}}/prebuilds
+      REALM_DISABLE_SYNC_TESTS: true
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:10]: Will this be removed before the merge?

File: CMakeLists.txt [hunk 9]
```diff
@@ -87,4 +87,5 @@ if(REALM_JS_BUILD_CORE_FROM_SOURCE)
     add_subdirectory(vendor/realm-core EXCLUDE_FROM_ALL)
 endif()
 
-add_subdirectory(src)
\ No newline at end of file
+add_subdirectory(src)
+add_definitions(-DNAPI_VERSION=3)
```
Comment (kraenhansen) [inline CMakeLists.txt:91]: Could this be moved somewhere more Node.js specific?

File: scripts/build-android.js [hunk 10]
```diff
@@ -51,24 +51,24 @@ if (options.arch) {
 
 const buildType = options.buildType;
 
-const ndkPath = process.env["ANDROID_NDK"];
+const ndkPath = process.env["ANDROID_NDK"] || process.env["ANDROID_NDK_HOME"];
```
Comment (kraenhansen) [inline scripts/build-android.js:54]: Why is this needed?

File: scripts/build-android.js [hunk 11]
```diff
@@ -51,24 +51,24 @@ if (options.arch) {
...
-  throw Error("ANDROID_NDK environment variable not set");
+  throw Error("ANDROID_NDK / ANDROID_NDK_HOME environment variable not set");
 }
 
 const sdkPath = getAndroidSdkPath();
 const cmakePath = getCmakePath(sdkPath);
 const cmakeVersion = getCmakeVersion(sdkPath);
 
-const buildPath = path.resolve(process.cwd(), "build-realm-android");
+const buildPath = path.resolve(process.cwd(), "prebuilds", "build-realm-android");
```
Comment (kraenhansen) [inline scripts/build-android.js:63]: Since `prebuilds` are specifically referring to the artifacts produced by the [prebuild](https://www.npmjs.com/package/prebuild) package, I'd say we should leave the android artifacts outside of that directory.

File: scripts/build-ios.sh [hunk 12]
```diff
@@ -81,6 +81,9 @@ for platform in "${PLATFORMS[@]}"; do
     esac
 done
 
+mkdir -p "$PROJECT_ROOT"/prebuilds
+test -e "$PROJECT_ROOT"/prebuilds/realm-js-ios.xcframework && cp -a "$PROJECT_ROOT"/prebuilds/realm-js-ios.xcframework
```
Comment (kraenhansen) [inline scripts/build-ios.sh:85]: Same with Android, I would appreciate if we didn't use the `prebuilds` directory to store the ios xcframework artifact.

Comment (kraenhansen) [reply scripts/build-ios.sh:85]: Also, why are we doing this in the first place? Won't this simply get overwritten in the subsequent steps of this script?

File: scripts/ci-scripts/ci-eslint.sh [hunk 13]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+npm ci --ignore-scripts
```
Comment (kraenhansen) [inline scripts/ci-scripts/ci-eslint.sh:3]: This will make it impractical to run this script locally. I'd suggest moving this to the invoking workflow instead.

```suggestion
```

File: scripts/ci-scripts/realm-example.sh [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+pushd examples/ReactExample
+npm ci
```
Comment (kraenhansen) [inline scripts/ci-scripts/realm-example.sh:3]: I suggest moving this to the invoking workflow.

```suggestion
```

File: scripts/ci-scripts/react-tests.sh [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+
+echo "building iOS binaries"
+#./scripts/build-ios.sh -s -c $CONFIGURATION
+
+# set_nvm_default
+# start_server
+
+pushd tests/react-test-app
+echo "tests/react-test-app npm ci --no-optional"
+npm ci --no-optional
```
Comment (kraenhansen) [inline scripts/ci-scripts/react-tests.sh:168]: I suggest moving this to the invoking workflow.

```suggestion
```

File: scripts/ci-scripts/react-tests.sh [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+echo "installing install-local"
+./node_modules/.bin/install-local
+# open_chrome
+# start_packager
+
+pushd ios
+pod install
```
Comment (kraenhansen) [inline scripts/ci-scripts/react-tests.sh:175]: I suggest moving this to the invoking workflow.

```suggestion
```

File: scripts/ci-scripts/react-tests.sh [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+#npm ci --ignore-scripts
+#npm run check-environment
+
+echo "building iOS binaries"
+#./scripts/build-ios.sh -s -c $CONFIGURATION
+
+# set_nvm_default
+# start_server
```
Comment (kraenhansen) [inline scripts/ci-scripts/react-tests.sh:164]: Is this left commented out to enable local debugging? It might make sense to leave it in and skip it if a `CI` variable is present.

File: scripts/ci-scripts/test-android.sh [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+pushd tests/react-test-app
+echo "installing react-test-app dependencies"
+npm ci --no-optional
+
+echo "installing manually packed realm package"
+npm install --save-optional  --ignore-scripts ../../realm.tgz
+
+echo "installing manually packed realm tests package"
+npm install --save-optional ../js/realm-tests.tgz
+
```
Comment (kraenhansen) [inline scripts/ci-scripts/test-android.sh:47]: Could all of these be combined into a single `npm install` invocation to save some time?

File: scripts/build-android.js (hunk 10 above)
Comment (takameyer) [reply scripts/build-android.js:54]: The NDK path in github actions is stored in ANDROID_NDK_HOME

File: CMakeLists.txt (hunk 9 above)
Comment (kneth) [reply CMakeLists.txt:91]: `dependencies.list` has `NAPI_VERSION=4`. See also https://github.com/realm/realm-js/blob/master/src/node/node_value.hpp#L77.

When node v12.x has end-of-life (April 2022), we will bump to NAPI version 5.

File: scripts/ci-scripts/test-node.sh [hunk 19]
```diff
@@ -0,0 +1,9 @@
+#/bin/bash
+SRCROOT=$(cd "$(dirname "$0")/../.." && pwd)
+npm run check-environment
+#start_server
```
Comment (kneth) [inline scripts/ci-scripts/test-node.sh:4]: So no sync tests?

**Comment by iamer** [comment]
superseded by #4262
